hack/verify-versions.sh

```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+SCRIPT_ROOT="$(dirname "${BASH_SOURCE}")/.."
+MANIFEST_ROOT="$SCRIPT_ROOT/deploy/single"
+FAILED=0
+
+for manifest in `ls ${MANIFEST_ROOT}`
+do
+		IMAGES=$(grep -oP "image: kong(\/kong-gateway)?:[\d\.]+" $MANIFEST_ROOT/$manifest | sort | uniq)
```

Comment:
this grep is not robust: it makes assumptions about formatting of the yaml that can fail to match other valid representations. For example, if the `image` value is quoted.

It would be safer to use a tool like [`yq`](https://github.com/marketplace/actions/yq-portable-yaml-processor#go-get) that actually parse the YAML rather than regexp-scanning.

Reply:
Is there a yq query that will get all images in the file without enumerating each of their locations? The naive text version does have the benefit that it doesn't need to know where all the images are (and needing to know where all the images are was the original cause of this problem).

Kustomize already normalizes the output pretty aggressively, so these should be consistent. It does strip quotes from images:

```
14:57:23+0200 esenin $ git diff                                                
diff --git a/config/variants/postgres/migration.yaml b/config/variants/postgres/migration.yaml
index 5e2ec962..cbc28ed9 100644
--- a/config/variants/postgres/migration.yaml
+++ b/config/variants/postgres/migration.yaml
@@ -20,7 +20,7 @@ spec:
         command: [ "/bin/sh", "-c", "until nc -zv $KONG_PG_HOST $KONG_PG_PORT -w1; do echo 'waiting for db'; sleep 1; done" ]
       containers:
       - name: kong-migrations
-        image: kong:2.7
+        image: "kong:2.7"
         env:
         - name: KONG_PG_PASSWORD
           value: kong
14:57:43+0200 esenin $ kustomize build config/variants/postgres | grep "image:"
        image: kong:2.7
        image: kong/kubernetes-ingress-controller:2.1.1
        image: kong:2.7
        image: postgres:9.5
        image: kong:2.7
        image: busybox
```

Reply:
> Is there a yq query that will get all images in the file without enumerating each of their locations?

Below is a yq query I used that gets all images, but not without enumerating each of their locations.

```
yq eval-all '.spec.template.spec.containers[].image' *.yaml -N
yq eval-all '.spec.template.spec.initContainers[].image' *.yaml -N
```

I don't like making an implicit assumption that the set of image fields is unspecified, and we want magic to claim that any field called `image` anywhere in the schema (and no fields called anything but `image`) will contain image names that are of interest for us. This is a theoretical problem anyway, unlikely to surface anytime soon (if ever).

> Kustomize already normalizes the output pretty aggressively, so these should be consistent. It does strip quotes from images

In my experience, an assumption like this (that we accept yaml, but believe the format to be in a specific format printed by a specific tool) is a common source of bugs. Sure, the YAMLs come from kustomize today, but maybe we'll change the tool some day, or Kustomize will change what they do, or we make a manual intervention to the file.
I know, principles are frowned upon, but there's a [wikipedia page](https://en.wikipedia.org/wiki/Robustness_principle) about the general instance of this phenomenon.

My suggestion is to:
- instead of the guesswork of identifying what field in the schema is an image name and what isn't, call these out directly
- accept any valid YAML

The above suggestion with `yq` seems to achieve both. LMKWYT

Reply:
Was overthinking it and thought it'd be necessary on an individual basis for each document, rather than just using the paths within the template spec. See https://github.com/rainest/kubernetes-ingress-controller/runs/4728498736?check_suite_focus=true for new example.